File: 2058-find-the-minimum-and-maximum-number-of-nodes-between-critical-points/2058-find-the-minimum-and-maximum-number-of-nodes-between-critical-points.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def nodesBetweenCriticalPoints(self, head: Optional[ListNode]) -> List[int]:
        
        prev, curr = head, head.next
        result = [float('inf'), -1]
        index = 1
        first_index = -1
        last_occurrence = -1
        while curr.next:
            maxima = prev.val < curr.val > curr.next.val
            minima = prev.val > curr.val < curr.next.val
            if maxima or minima:
                if last_occurrence == -1:
                    first_index = index
                else:
                    result[0] = min(result[0], index - last_occurrence)
                    result[1] = index - first_index
                last_occurrence = index
            prev = curr
            curr = curr.next  
            index += 1

        return [-1, -1] if result[-1] == -1 else result

File: 2058-find-the-minimum-and-maximum-number-of-nodes-between-critical-points/2058-find-the-minimum-and-maximum-number-of-nodes-between-critical-points.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {number[]}
 */
var nodesBetweenCriticalPoints = function(head) {
    let prev = head;
    let curr = head.next;
    let first_occurrence = -1;
    let last_occurrence = -1
    let result = [Infinity, -Infinity];
    let index = 1;
    while (curr.next) {
        let maxima = prev.val < curr.val && curr.val > curr.next.val;  
        let minima = prev.val > curr.val && curr.val < curr.next.val;
        console.log(maxima, minima)
        if (maxima || minima) {
            console.log(index);
            if (last_occurrence === -1) {
                first_occurrence = index;
            } else {
                result[0] = Math.min(result[0], index - last_occurrence);
                result[1] = index - first_occurrence;
            }
            last_occurrence = index;
        }

        prev = curr
        curr = curr.next;
        index++;
    }

    return result[0] !== Infinity ? result : [-1, -1];
};